### PR TITLE
Implemented the loss NTK calculation

### PR DESCRIPTION
**Pull request by ma-sauter**

### CODE REVIEW

File: znnl/training_strategies/simple_training.py [hunk 1]
```diff
@@ -112,6 +112,7 @@ def __init__(
         # Add the loss and accuracy function to the recorders and re-instantiate them
```
Comment (KonstiNik) [inline znnl/training_strategies/simple_training.py:112]: adapt the comment

**Review by KonstiNik — CHANGES_REQUESTED** [review]
Good job in this PR. The first draft looks pretty nice. 

I have two major comments:
1. In general, a test should cover as many aspects of the code as possible. And a test should test the desired aspect as simply as possible. This includes trying to rely as little as possible on existing methods. E.g. including an existing data generator is not as good practice as creating some dummy test data. 
2. With your implementation, we would need to duplicate all observables for the loss ntk that we already have for the ntk. One can avoid this by having the option to either use a recorder for the loss ntk or the regular ntk. This could be done with one keyword at initialization e.g..

File: znnl/training_recording/jax_recording.py [hunk 2]
```diff
@@ -254,19 +271,45 @@ def instantiate_recorder(self, data_set: dict = None, overwrite: bool = False):
...
+                "eigenvalues" in self._selected_properties,
+                "trace" in self._selected_properties,
+            ]
+        ):
             self._compute_ntk = True
 
+        # Check if we need a loss NTK computation and update the class accordingly
+        if any(
+            [
+                "loss_ntk" in self._selected_properties,
```
[the remark this answers is not in the document]
Comment (ma-sauter) [reply znnl/training_recording/jax_recording.py:292]: I agree that there's room for improvement, but I think if we introduce a flag like this here we should maybe also discuss more changes to the recorder. We should talk about this in person or in a meeting, but I'd like to make sure that the tests are working first because it's more urgent for the DPG if that's fine.

File: znnl/training_strategies/simple_training.py (hunk 1 above)
Comment (ma-sauter) [reply znnl/training_strategies/simple_training.py:112]: I'm not sure what you mean, could you explain this further?

Comment (KonstiNik) [reply znnl/training_strategies/simple_training.py:112]: I mean you now also include the model there.

Comment (ma-sauter) [reply znnl/training_strategies/simple_training.py]: Was fixed in commit [d17e514](https://github.com/zincware/ZnNL/pull/109/commits/d17e5149e0e93df1ee89509ff16fca738950f430)

File: znnl/training_recording/jax_recording.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (ma-sauter) [reply znnl/training_recording/jax_recording.py]: The flag got introduced in commit [1dac434](https://github.com/zincware/ZnNL/pull/109/commits/1dac43435ff563c7e89d931906c752d0d4788722)

File: znnl/analysis/loss_ntk_calculation.py [hunk 3]
```diff
@@ -0,0 +1,233 @@
...
+from typing import Callable
+
+import jax
+import jax.numpy as np
+import neural_tangents as nt
+
+from znnl.models.jax_model import JaxModel
+
+
+class LossNTKCalculation:
```
[the remark this answers is not in the document]
Comment (ma-sauter) [reply znnl/analysis/loss_ntk_calculation.py:37]: Fixed in [b6ebb8d](https://github.com/zincware/ZnNL/pull/109/commits/b6ebb8d5f0b5991fb8f17e4fcbfe7403a6566fa5)

File: znnl/analysis/loss_ntk_calculation.py [hunk 4]
```diff
@@ -0,0 +1,233 @@
...
+        target: np.ndarray
+                The unshaped target.
+        """
+        return datapoint[:, :input_dimension].reshape(
+            batch_length, *input_shape[1:]
+        ), datapoint[:, input_dimension:].reshape(batch_length, *target_shape[1:])
+
+    def _function_for_loss_ntk(self, params, datapoint) -> float:
+        """
+        Helper function to create a subloss apply function.
```
[the remark this answers is not in the document]
Comment (ma-sauter) [reply znnl/analysis/loss_ntk_calculation.py]: Changed in [b6ebb8d](https://github.com/zincware/ZnNL/pull/109/commits/b6ebb8d5f0b5991fb8f17e4fcbfe7403a6566fa5)

File: CI/unit_tests/analysis/test_loss_ntk_calculation.py [hunk 5]
```diff
@@ -0,0 +1,259 @@
...
+        }
+
+        # Initialize the loss NTK calculation
+        loss_ntk_calculator = LossNTKCalculation(
+            metric_fn=LPNorm(order=2),
+            model=production_model,
+            dataset=data_set,
+        )
+
+        # Setup a test dataset for reshaping
```
[the remark this answers is not in the document]
Comment (ma-sauter) [reply CI/unit_tests/analysis/test_loss_ntk_calculation.py:85]: Changed in [b6ebb8d](https://github.com/zincware/ZnNL/pull/109/commits/b6ebb8d5f0b5991fb8f17e4fcbfe7403a6566fa5)

File: CI/unit_tests/analysis/test_loss_ntk_calculation.py [hunk 6]
```diff
@@ -0,0 +1,259 @@
...
+            test_data_set["inputs"][0],
+        )
+        subloss = np.linalg.norm(applied_model - test_data_set["targets"][0], ord=2)
+
+        # Check that the two losses are the same
+        assert subloss - subloss_from_NTK < 1e-5
+
+    def test_loss_NTK_calculation(self):
+        """
+        Test the Loss NTK calculation.
```
[the remark this answers is not in the document]
Comment (ma-sauter) [reply CI/unit_tests/analysis/test_loss_ntk_calculation.py]: I changed this to random data with a smaller feed forward model in [b6ebb8d](https://github.com/zincware/ZnNL/pull/109/commits/b6ebb8d5f0b5991fb8f17e4fcbfe7403a6566fa5)

File: CI/unit_tests/analysis/test_loss_ntk_calculation.py [hunk 7]
```diff
@@ -0,0 +1,259 @@
...
+        loss_ntk = loss_ntk_calculator.compute_loss_ntk(
+            x_i=data_set, model=production_model
+        )["empirical"]
+
+        # Now for comparison calculate regular ntk
+        ntk = production_model.compute_ntk(data_set["inputs"], infinite=False)[
+            "empirical"
+        ]
+        # Calculate Loss derivative fn
+        loss_derivative_calculator = LossDerivative(LPNormLoss(order=2))
```
[the remark this answers is not in the document]
Comment (ma-sauter) [reply CI/unit_tests/analysis/test_loss_ntk_calculation.py]: Was changed in [b6ebb8d](https://github.com/zincware/ZnNL/pull/109/commits/b6ebb8d5f0b5991fb8f17e4fcbfe7403a6566fa5)

File: CI/unit_tests/analysis/test_loss_ntk_calculation.py [hunk 8]
```diff
@@ -0,0 +1,259 @@
...
+        # note: here we need the derivatives of the subloss, not the regular loss fn
+        loss_derivatives = onp.empty(shape=(len(predictions), len(predictions[0])))
+        for i in range(len(loss_derivatives)):
+            # The weird indexing here is because of axis constraints in LPNormLoss
+            loss_derivatives[i] = loss_derivative_calculator.calculate(
+                predictions[i : i + 1], data_set["targets"][i : i + 1]
+            )[0]
+
+        # Calculate the loss NTK from the loss derivatives and the ntk
+        loss_ntk_2 = np.einsum(
```
[the remark this answers is not in the document]
Comment (ma-sauter) [reply CI/unit_tests/analysis/test_loss_ntk_calculation.py:227]: Yes, at first I implemented it as a simple sum which I was sure to be correct and only introduced the einsum afterwards while making sure that the results match with the previous implementation.

File: znnl/analysis/loss_ntk_calculation.py [hunk 9]
```diff
@@ -0,0 +1,241 @@
...
+        metric_fn: Callable,
+        model: JaxModel,
+        dataset: dict,
+    ):
+        """
+        Constructor for the loss ntk calculation class.
+
+        Parameters
+        ----------
+
```
Comment (SamTov) [inline znnl/analysis/loss_ntk_calculation.py:60]: Can you remove this space

File: znnl/analysis/loss_ntk_calculation.py [hunk 10]
```diff
@@ -0,0 +1,241 @@
...
+                trace_axes=(),
+                vmap_axes=0,
+            ),
+            batch_size=self.ntk_batch_size,
+            store_on_device=self.store_on_device,
+        )
+        self.empirical_ntk_jit = jax.jit(empirical_ntk)
+
+    @staticmethod
+    def _reshape_dataset(dataset):
```
Comment (SamTov) [inline znnl/analysis/loss_ntk_calculation.py:98]: You can remove these from the class and put them into a utils file.

File: znnl/analysis/loss_ntk_calculation.py [hunk 11]
```diff
@@ -0,0 +1,241 @@
...
+        return np.concatenate(
+            (
+                dataset["inputs"].reshape(dataset["inputs"].shape[0], -1),
+                dataset["targets"].reshape(dataset["targets"].shape[0], -1),
+            ),
+            axis=1,
+        )
+
+    @staticmethod
+    def _unshape_data(
```
Comment (SamTov) [inline znnl/analysis/loss_ntk_calculation.py:122]: Sam here.

File: znnl/analysis/loss_ntk_calculation.py [hunk 12]
```diff
@@ -0,0 +1,241 @@
...
+        model : JaxModel
+                The model for which to calculate the loss NTK.
+
+        dataset : dict
+                The dataset for which to calculate the loss NTK.
+                The dictionary should contain the keys "inputs" and "targets".
+        """
+
+        # Set the attributes
+        self.ntk_batch_size = model.ntk_batch_size
```
Comment (SamTov) [inline znnl/analysis/loss_ntk_calculation.py:76]: It might be better to make all of these arguments in this calculation. Especially when we later move into the new measurement system, this will all need to be self contained. Things like `store_on_device` are only pertinent to this calculator.

File: znnl/analysis/loss_ntk_calculation.py [hunk 13]
```diff
@@ -0,0 +1,241 @@
...
+                The shape of the original input.
+        target_shape : tuple
+                The shape of the original target.
+        batch_length : int
+                The length of the batch, i.e. the number of actual datapoints
+                inside the datapoint variable
+
+        Returns
+        -------
+        input: np.ndarray
```
Comment (SamTov) [inline znnl/analysis/loss_ntk_calculation.py:148]: Can you add some shape information here? It can be (batch * input size, ) or anything, but just some information about what I will get back. What you mean by unshape is also very unclear. Is it flattening is it reshaping, unshape doesn't have a real meaning.

File: znnl/analysis/loss_ntk_calculation.py [hunk 14]
```diff
@@ -0,0 +1,241 @@
...
+                The unshaped input.
+        target: np.ndarray
+                The unshaped target.
+        """
+        return datapoint[:, :input_dimension].reshape(
+            batch_length, *input_shape[1:]
+        ), datapoint[:, input_dimension:].reshape(batch_length, *target_shape[1:])
+
+    def _function_for_loss_ntk(self, params, datapoint) -> float:
```
Comment (SamTov) [inline znnl/analysis/loss_ntk_calculation.py:157]: I would prefer different naming here. Is it an apply function on flattened data, a loss function. What do you mean by subloss? Loss between two data points is just loss. Function for loss ntk could be anything.

File: znnl/visualization/tsne_visualizer.py
Comment (SamTov) [inline znnl/visualization/tsne_visualizer.py:1]: If the notebook is not clear, can you clear it of outputs.

**Review by SamTov — CHANGES_REQUESTED** [review]
I have a few comments. If you go through and address them all I can go back over it but in general, I like it and am happy to have it merged soon.
